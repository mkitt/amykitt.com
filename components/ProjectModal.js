// @flow
import React from 'react'
import css, { media3 } from '../styles/css'
import Dismiss from './Dismiss'
import Markdown from './Markdown'
import ProjectPaddles from './ProjectPaddles'
import View from './View'
import type { Project } from '../types/app.js.flow'

type Props = {
  onCloseModal: Function,
  onNextDetail: Function,
  onPrevDetail: Function,
  project: Project | null | void,
}

const style = css({
  position: 'fixed',
  top: 0,
  right: 0,
  bottom: 0,
  left: 0,
  zIndex: 20,
  transition: 'opacity 0.4s ease, background-color 0.4s ease',
})

const inactiveStyle = css({
  opacity: 0,
  pointerEvents: 'none',
  backgroundColor: 'transparent',
})

const activeStyle = css({
  opacity: 1,
  WebkitOverflowScrolling: 'touch',
  overflowY: 'auto',
  backgroundColor: 'rgba(124, 124, 129, 0.5)',
})

const markdownStyle = css(
  {
    position: 'relative',
    maxWidth: 1292,
    minHeight: 'calc(100vh - 160px)',
    paddingTop: 15,
    paddingRight: 20,
    paddingBottom: 20,
    paddingLeft: 20,
    marginTop: 160,
    marginRight: 'auto',
    marginLeft: 'auto',
    backgroundColor: '#fff',
    border: '5px solid #fff',
  },
  media3({
    paddingRight: 80,
    paddingBottom: 80,
    paddingLeft: 80,
  }),
)

let modalElement = null

export default class extends React.PureComponent {
  props: Props

  componentDidUpdate() {
    const { project } = this.props
    if (!(document && document.body)) { return }
    if (project) {
      document.body.classList.add('isModalActive')
    } else if (!project) {
      modalElement = null
      document.body.classList.remove('isModalActive')
    }
    if (modalElement) {
      modalElement.scrollTop = 0
    }
  }

  onClickModal = (e: any) => {
    const classList = e.target.classList
    if (classList.contains('Modal') || classList.contains('jsDismissModal')) {
      this.props.onCloseModal()
    }
  }

  render() {
    const { onNextDetail, onPrevDetail, project } = this.props
    if (!project) {
      return <View className={`Modal ${style} ${inactiveStyle}`} />
    }
    return (
      <View
        className={`Modal ${style} ${activeStyle}`}
        onClick={this.onClickModal}
        ref={(el) => { modalElement = el }}
      >
        <View className={markdownStyle}>
          <Markdown
            source={project.body}
          />
          <ProjectPaddles
            onNextDetail={onNextDetail}
            onPrevDetail={onPrevDetail}
          />
        </View>
        <Dismiss className="jsDismissModal" />
      </View>
    )
  }
}
