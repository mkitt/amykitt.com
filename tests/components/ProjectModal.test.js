import React from 'react'
import renderer from 'react-test-renderer'
import ProjectModal from '../../components/ProjectModal'

const noop = () => {}

describe('ProjectModal', () => {
  let component

  beforeEach(() => {
    const project = {
      body: 'Project 1 _body_',
      id: 'project-1',
      tile: 'project-1-tiles',
      tileWidth: 600,
      tileHeight: 400,
      title: 'Project 1 title',
    }

    component = (
      <ProjectModal
        onCloseModal={noop}
        onNextDetail={noop}
        onPrevDetail={noop}
        project={project}
      />
    )
  })

  it('shows the correct snapshot for the ProjectModal component tree', () => {
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
