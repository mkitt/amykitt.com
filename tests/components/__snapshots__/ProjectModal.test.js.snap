// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectModal shows the correct snapshot for the ProjectModal component tree 1`] = `
<div
  className="Modal css-1y45o1n css-1bpvw5n"
  onClick={[Function]}
>
  <div
    className={
      Object {
        "data-css-1x9ll69": "",
      }
    }
  >
    <div
      className="css-jlylsu"
    >
      <p>
        Project 1 
        <em>
          body
        </em>
      </p>
    </div>
    <div
      className={
        Object {
          "data-css-1da2mo1": "",
        }
      }
    >
      <button
        className="css-1fqszn3 css-12ml4zg"
        onClick={[Function]}
        type="button"
      >
        Prev
      </button>
      <span
        className={
          Object {
            "data-css-1qvhb5d": "",
          }
        }
      >
         / 
      </span>
      <button
        className="css-1fqszn3 css-12ml4zg"
        onClick={[Function]}
        type="button"
      >
        Next
      </button>
    </div>
  </div>
  <button
    className="css-1fqszn3 jsDismissModal css-rh1ike"
    title="Dismiss"
    type="button"
  >
    <svg
      aria-hidden={true}
      height={24}
      viewBox="0 0 24 24"
      width={24}
    >
      <g
        stroke="currentColor"
      >
        <line
          x1="2"
          x2="22"
          y1="2"
          y2="22"
        />
        <line
          x1="22"
          x2="2"
          y1="2"
          y2="22"
        />
      </g>
    </svg>
  </button>
</div>
`;
